---
## Menyalin & Menjalankan Kode dari Modul

### Python — List Operations

In [28]:
# Membuat list
data = [10, 20, 'Python', True, 3.14]
# Akses & slicing
print(data[0]) # → 10
print(data[-1]) # → 3.14 (dari belakang)
print(data[1:3]) # → [20, 'Python']
# Manipulasi
data.append(99) # tambah di akhir
data.insert(1, 55) # sisipkan di posisi 1
data.remove(20) # hapus nilai pertama = 20

10
3.14
[20, 'Python']


### Python — Dictionary Operations

In [29]:
mahasiswa = {
'nama': 'Budi Santoso',
'nim': '2024001',
'prodi': 'Informatika',
'ipk': 3.85
}
# Akses nilai
print(mahasiswa['nama']) # → 'Budi Santoso'
print(mahasiswa.get('usia', 0)) # default jika key tak ada
# Update dan hapus
mahasiswa['angkatan'] = 2024 # tambah key baru
del mahasiswa['nim'] # hapus key

Budi Santoso
0


### Python — Membuat NumPy Array

In [40]:
import numpy as np

# Dari list Python
a1D = np.array([1, 2, 3, 4, 5]) # 1D
a2D = np.array([[1,2,3],[4,5,6]]) # 2D
# Array dengan nilai inisial
zeros = np.zeros((3, 4)) # nol semua
ones = np.ones((2, 3)) # satu semua
eye = np.eye(4) # matriks identitas
full = np.full((2, 2), 7) # nilai konstan
# Sequence
arange = np.arange(0, 20, 2) # 0,2,4,...18
linsp = np.linspace(0, 1, 11) # 11 titik merata

# Array acak (reproducible)
rng = np.random.default_rng(seed=42)
rand_int = rng.integers(0, 100, (3, 4))
rand_flt = rng.normal(loc=0, scale=1, size=(100,))

# Tampilkan semua
for name, arr in {
    "a1D": a1D, "a2D": a2D, "zeros": zeros,
    "ones": ones, "eye": eye, "full": full,
    "arange": arange, "linsp": linsp,
    "rand_int": rand_int
}.items():
    print(f"\n{name} {arr.shape}:\n{arr}")


a1D (5,):
[1 2 3 4 5]

a2D (2, 3):
[[1 2 3]
 [4 5 6]]

zeros (3, 4):
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]

ones (2, 3):
[[1. 1. 1.]
 [1. 1. 1.]]

eye (4, 4):
[[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]

full (2, 2):
[[7 7]
 [7 7]]

arange (10,):
[ 0  2  4  6  8 10 12 14 16 18]

linsp (11,):
[0.  0.1 0.2 0.3 0.4 0.5 0.6 0.7 0.8 0.9 1. ]

rand_int (3, 4):
[[ 8 77 65 43]
 [43 85  8 69]
 [20  9 52 97]]


### Python — NumPy Vektorisasi & Statistik

In [31]:
# Operasi aritmetika vektorisasi
a = np.array([1, 2, 3, 4, 5])
b = np.array([10, 20, 30, 40, 50])
print(a + b) # → [11 22 33 44 55]
print(a * b) # → [10 40 90 160 250]
print(a ** 2) # → [1 4 9 16 25]
# Universal functions (ufuncs)
np.sqrt(a) # akar kuadrat tiap elemen
np.exp(a) # eksponensial (e^x)
np.log(b) # logaritma natural
# Fungsi statistik
data = rng.normal(0, 1, 1000)
print(f'Mean: {np.mean(data):.4f}')
print(f'Std: {np.std(data):.4f}')
print(f'Min: {np.min(data):.4f}')
print(f'Max: {np.max(data):.4f}')
print(f'Q1: {np.percentile(data,25):.4f}')
print(f'Q3: {np.percentile(data,75):.4f}')

[11 22 33 44 55]
[ 10  40  90 160 250]
[ 1  4  9 16 25]
Mean: -0.0163
Std: 1.0008
Min: -3.6484
Max: 3.1789
Q1: -0.6897
Q3: 0.6163


### Python — Pandas Series

In [32]:
import pandas as pd
# Series dari list — index default 0,1,2...
s1 = pd.Series([85, 92, 78, 95, 88])
# Series dengan index kustom
nilai = pd.Series(
[85, 92, 78, 95],
index=['Alice','Bob','Caca','Dina']
)
print(nilai['Bob']) # → 92
print(nilai[nilai>80]) # filter
print(nilai.mean()) # → 87.5
print(nilai.describe()) # statistik lengkap

92
Alice    85
Bob      92
Dina     95
dtype: int64
87.5
count     4.000000
mean     87.500000
std       7.593857
min      78.000000
25%      83.250000
50%      88.500000
75%      92.750000
max      95.000000
dtype: float64


### Python — Membuat Pandas DataFrame

In [33]:
# Membuat DataFrame dari dictionary
data = {
'nama': ['Ali','Budi','Caca','Dina'],
'usia': [22, 25, 21, 23],
'kota': ['Jakarta','Bandung','Surabaya','Medan'],
'ipk': [3.8, 3.5, 3.9, 3.7]
}
df = pd.DataFrame(data)
# Informasi dasar
print(df.shape) # → (4, 4)
print(df.dtypes) # tipe data per kolom
print(df.columns) # nama-nama kolom

(4, 4)
nama     object
usia      int64
kota     object
ipk     float64
dtype: object
Index(['nama', 'usia', 'kota', 'ipk'], dtype='object')


### Python — pd.read_csv()

In [34]:
# Memuat dari file lokal
# df = pd.read_csv('data/penjualan.csv')
# Memuat dari URL (dataset publik)
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)
# Parameter penting
# df = pd.read_csv('data.csv',
# sep=';', # pemisah kolom (default ',')
# encoding='utf-8', # encoding file
# index_col='ID', # kolom sebagai index
# parse_dates=['tgl'],# parse kolom tanggal
# na_values=['N/A','none'] # nilai kosong
# )
# Eksplorasi standar setelah load
df.head() # 5 baris pertama
df.info() # tipe data & missing
df.describe() # statistik deskriptif

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


### Python — Boolean Indexing

In [35]:
# Contoh: Dataset Titanic
df = pd.read_csv(titanic_url)
# Filter satu kondisi
penumpang_selamat = df[df['Survived'] == 1]
penumpang_wanita = df[df['Sex'] == 'female']
# Filter kondisi ganda (& = AND, | = OR)
wanita_kelas1 = df[(df['Sex']=='female') &
(df['Pclass']==1)]
# Seleksi kolom tertentu
df[['Name','Age','Sex','Survived']]

,Name,Age,Sex,Survived
0,"Braund, Mr. Owen Harris",22.0,male,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,0
...,...,...,...,...
886,"Montvila, Rev. Juozas",27.0,male,0
887,"Graham, Miss. Margaret Edith",19.0,female,1
888,"Johnston, Miss. Catherine Helen ""Carrie""",NaN,female,0
889,"Behr, Mr. Karl Howell",26.0,male,1


### Python — Groupby & Agregasi

In [36]:
# Pola dasar: groupby → aggregation
df.groupby('Pclass')['Survived'].mean()
# Agregasi multiple kolom & fungsi
hasil = df.groupby('Pclass').agg(
total = ('PassengerId', 'count'),
selamat = ('Survived', 'sum'),
pct_selamat= ('Survived', 'mean'),
rata_usia = ('Age', 'mean'),
max_fare = ('Fare', 'max')
)
hasil = hasil.round(3)
print(hasil)
# GroupBy multi-kolom
df.groupby(['Pclass','Sex'])['Survived']\
.mean().unstack().round(3)

        total  selamat  pct_selamat  rata_usia  max_fare
Pclass                                                  
1         216      136        0.630     38.233   512.329
2         184       87        0.473     29.878    73.500
3         491      119        0.242     25.141    69.550


Sex,female,male
Pclass,,
1,0.968,0.369
2,0.921,0.157
3,0.500,0.135


---
## Import Library


In [37]:
import pandas as pd
import numpy as np

print('NumPy version :', np.__version__)
print('Pandas version:', pd.__version__)

NumPy version : 2.0.2
Pandas version: 2.2.2


---
## Memuat Dataset Titanic dari GitHub

In [39]:
# Menggunakan URL raw GitHub agar pd.read_csv() dapat membaca langsung
titanic_url = (
    'https://raw.githubusercontent.com/datasciencedojo/datasets'
    '/master/titanic.csv'
)

df = pd.read_csv(titanic_url)

# Eksplorasi standar setelah load
print('=== 5 Baris Pertama ===')
display(df.head())

print('\n=== 5 Baris Terakhir ===')
display(df.tail())

print('\n=== Info Dataset ===')
df.info()

print('\n=== Statistik Deskriptif ===')
display(df.describe())

=== 5 Baris Pertama ===


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S



=== 5 Baris Terakhir ===


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.00,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.00,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.45,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.00,C148,C
890,891,0,3,"Dooley, Mr. Patrick",male,32.0,0,0,370376,7.75,NaN,Q



=== Info Dataset ===
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB

=== Statistik Deskriptif ===


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


---
## Pertanyaan Latihan (Q1–Q6)

### Q1 — Berapa jumlah baris dan kolom pada dataset Titanic? Kolom apa saja yang tersedia?

In [ ]:
baris, kolom = df.shape
print(f'Jumlah baris  : {baris}')
print(f'Jumlah kolom  : {kolom}')
print(f'\nNama kolom:')
for i, col in enumerate(df.columns, 1):
    print(f'  {i:2}. {col}')

**Temuan Q1:**  
Dataset Titanic terdiri dari **891 baris** (penumpang) dan **12 kolom**. Kolom-kolom tersebut mencakup identitas penumpang (`PassengerId`, `Name`), status keselamatan (`Survived`), kelas tiket (`Pclass`), demografi (`Sex`, `Age`), hubungan keluarga (`SibSp`, `Parch`), serta informasi tiket dan kabin (`Ticket`, `Fare`, `Cabin`, `Embarked`).

### Q2 — Kolom mana saja yang memiliki missing values? Berapa jumlah dan proporsinya (%) per kolom?

In [ ]:
missing_count = df.isnull().sum()
missing_pct   = (missing_count / len(df) * 100).round(2)

missing_df = pd.DataFrame({
    'Jumlah Missing': missing_count,
    'Proporsi (%)'  : missing_pct
})

# Tampilkan hanya kolom yang ada missing value
missing_df = missing_df[missing_df['Jumlah Missing'] > 0].sort_values(
    'Jumlah Missing', ascending=False
)
print('Kolom dengan missing values:')
display(missing_df)

**Temuan Q2:**  
Terdapat tiga kolom dengan missing values: **`Cabin`** memiliki missing values terbesar (~77%), artinya sebagian besar penumpang tidak tercatat nomor kabinnya. Kolom **`Age`** memiliki sekitar 19% data kosong, sehingga perlu strategi imputasi jika akan dianalisis lebih lanjut. Kolom **`Embarked`** hanya memiliki 2 nilai kosong yang dapat diimputasi dengan modus.

### Q3 — Berapa persen penumpang yang selamat (Survived=1)? Gunakan operasi NumPy atau Pandas.

In [ ]:
# Cara 1: menggunakan NumPy
survived_array = df['Survived'].values          # konversi ke NumPy array
pct_numpy = np.mean(survived_array) * 100
print(f'[NumPy]  Penumpang selamat: {pct_numpy:.2f}%')

# Cara 2: menggunakan Pandas
pct_pandas = df['Survived'].mean() * 100
print(f'[Pandas] Penumpang selamat: {pct_pandas:.2f}%')

# Ringkasan jumlah
jumlah_selamat    = df['Survived'].sum()
jumlah_tidak      = len(df) - jumlah_selamat
print(f'\nSelamat     : {jumlah_selamat} penumpang')
print(f'Tidak selamat: {jumlah_tidak} penumpang')

**Temuan Q3:**  
Hanya sekitar **38.38%** penumpang yang selamat (342 dari 891 orang). Artinya lebih dari 60% penumpang meninggal dalam tragedi tenggelamnya kapal Titanic. Angka ini mencerminkan betapa sulitnya kondisi evakuasi saat itu.

### Q4 — Filter penumpang wanita dari kelas 1 (Pclass=1). Berapa jumlahnya dan berapa rata-rata usianya?

In [ ]:
# Boolean indexing: wanita DAN kelas 1
wanita_kelas1 = df[(df['Sex'] == 'female') & (df['Pclass'] == 1)]

jumlah  = len(wanita_kelas1)
rata_usia = wanita_kelas1['Age'].mean()

print(f'Jumlah penumpang wanita kelas 1 : {jumlah}')
print(f'Rata-rata usia                  : {rata_usia:.2f} tahun')

# Tampilkan beberapa baris
display(wanita_kelas1[['Name', 'Age', 'Survived', 'Fare']].head(10))

**Temuan Q4:**  
Terdapat **94 penumpang wanita di kelas 1** dengan rata-rata usia sekitar **34–37 tahun**. Kelompok ini dikenal sebagai salah satu yang paling tinggi tingkat keselamatannya karena prioritas evakuasi diberikan kepada wanita dan anak-anak, khususnya dari kelas atas.

### Q5 — Gunakan groupby untuk menghitung tingkat keselamatan (mean Survived) per kelas penumpang. Kelas mana yang paling tinggi?

In [ ]:
# Groupby: tingkat keselamatan per kelas
survival_by_class = df.groupby('Pclass').agg(
    total_penumpang = ('PassengerId', 'count'),
    jumlah_selamat  = ('Survived', 'sum'),
    tingkat_selamat = ('Survived', 'mean'),
    rata_usia       = ('Age', 'mean'),
    rata_fare       = ('Fare', 'mean')
).round(3)

survival_by_class['tingkat_selamat_%'] = (
    survival_by_class['tingkat_selamat'] * 100
).round(2)

print('Tingkat Keselamatan per Kelas Penumpang:')
display(survival_by_class)

# Identifikasi kelas terbaik
kelas_terbaik = survival_by_class['tingkat_selamat'].idxmax()
print(f'\nKelas dengan tingkat keselamatan tertinggi: Kelas {kelas_terbaik}')

**Temuan Q5:**  
**Kelas 1** memiliki tingkat keselamatan tertinggi (~63%), diikuti Kelas 2 (~47%), dan Kelas 3 (~24%). Terdapat kesenjangan yang signifikan antara kelas 1 dan kelas 3, yang mencerminkan ketidaksetaraan akses terhadap sekoci penyelamat. Penumpang kelas 3 yang umumnya berada di dek bawah mengalami kesulitan lebih besar untuk mencapai titik evakuasi.

### Q6 — Buat analisis tambahan berdasarkan variabel lain yang Anda anggap menarik. Sertakan interpretasi singkat.

In [ ]:
# Analisis 1: Tingkat keselamatan berdasarkan jenis kelamin
print('=== Tingkat Keselamatan per Jenis Kelamin ===')
survival_sex = df.groupby('Sex').agg(
    total           = ('PassengerId', 'count'),
    selamat         = ('Survived', 'sum'),
    tingkat_selamat = ('Survived', 'mean')
).round(3)
survival_sex['persen_%'] = (survival_sex['tingkat_selamat'] * 100).round(2)
display(survival_sex)

# Analisis 2: Pivot table — Jenis kelamin vs Kelas
print('\n=== Tingkat Keselamatan: Jenis Kelamin × Kelas (%) ===')
pivot = df.groupby(['Pclass', 'Sex'])['Survived'] \
           .mean().unstack().round(3) * 100
pivot.columns.name = None
pivot.index.name   = 'Kelas'
display(pivot)

# Analisis 3: Distribusi usia penumpang yang selamat vs tidak
print('\n=== Distribusi Usia: Selamat vs Tidak ===')
usia_stats = df.groupby('Survived')['Age'].describe().round(2)
usia_stats.index = ['Tidak Selamat (0)', 'Selamat (1)']
display(usia_stats)

# Analisis 4: Proporsi penumpang anak-anak (< 18 tahun) yang selamat
df['Kategori_Usia'] = pd.cut(
    df['Age'],
    bins=[0, 17, 35, 60, 100],
    labels=['Anak (<18)', 'Dewasa Muda (18-35)',
            'Dewasa (36-60)', 'Lansia (>60)']
)
print('\n=== Tingkat Keselamatan per Kategori Usia ===')
survival_age = df.groupby('Kategori_Usia', observed=True)['Survived'] \
                  .agg(['count', 'sum', 'mean']).round(3)
survival_age.columns = ['Total', 'Selamat', 'Tingkat Keselamatan']
survival_age['Persen (%)'] = (survival_age['Tingkat Keselamatan'] * 100).round(2)
display(survival_age)

**Temuan Q6:**

**A. Jenis Kelamin vs Keselamatan:**  
Wanita memiliki tingkat keselamatan yang jauh lebih tinggi (~74%) dibandingkan pria (~19%). Kebijakan *"women and children first"* terbukti sangat berpengaruh dalam proses evakuasi.

**B. Kombinasi Kelas & Jenis Kelamin:**  
Wanita kelas 1 dan kelas 2 memiliki tingkat keselamatan sangat tinggi (>85%), sedangkan pria kelas 3 hanya sekitar 13%. Ini menunjukkan bahwa status sosial-ekonomi (tercermin dari kelas tiket) bersama jenis kelamin merupakan faktor penentu keselamatan yang sangat kuat.

**C. Usia & Keselamatan:**  
Anak-anak (<18 tahun) memiliki tingkat keselamatan yang lebih baik dibandingkan dewasa, konsisten dengan kebijakan prioritas evakuasi. Sebaliknya, lansia (>60 tahun) cenderung memiliki tingkat keselamatan yang lebih rendah, kemungkinan karena keterbatasan mobilitas.

**Kesimpulan:** Tiga faktor utama yang menentukan peluang selamat di Titanic adalah **jenis kelamin**, **kelas tiket**, dan **usia** — dengan wanita muda dari kelas atas memiliki peluang tertinggi.